libkbfs/file_data.go

```diff
@@ -520,44 +543,294 @@ func (fd *fileData) createIndirectBlock(
-	if err != nil {
-		return err
+		fd.log.CDebugf(ctx, "New right block for file %d, level %d, ptr %v",
+			fd.rootBlockPointer(), i, newPtr)
+
+		pblock.IPtrs = append(pblock.IPtrs, IndirectFilePtr{
+			BlockInfo: BlockInfo{
+				BlockPointer: newPtr,
+				EncodedSize:  0,
+			},
+			Off: off,
+		})
+		rightParentBlocks[i].pblock = pblock
+		rightParentBlocks[i].childIndex = len(pblock.IPtrs) - 1
+
+		rblock := &FileBlock{}
+		if i != len(parentBlocks)-1 {
+			rblock.IsInd = true
+			pblock = rblock
+		}
+
+		err = fd.cacher(newPtr, rblock)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		newDirtyPtrs = append(newDirtyPtrs, newPtr)
 	}
-	err = fd.cacher(ptr, pblock)
-	if err != nil {
-		return err
+
+	// All parents up to and including the lowest ancestor with room
+	// will have to change, so mark them as dirty.
+	ptr := fd.rootBlockPointer()
+	for i := 0; i <= lowestAncestorWithRoom; i++ {
+		pb := parentBlocks[i]
+		if err := fd.cacher(ptr, pb.pblock); err != nil {
+			return nil, nil, err
+		}
+		newDirtyPtrs = append(newDirtyPtrs, ptr)
+		ptr = pb.childIPtr().BlockPointer
+		rightParentBlocks[i].pblock = pb.pblock
+		rightParentBlocks[i].childIndex = len(pb.pblock.IPtrs) - 1
 	}
-	return nil
+
+	return rightParentBlocks, newDirtyPtrs, nil
+}
+
+// shiftBlocksToFillHole should be called after newRightBlock when the
+// offset for the new block (`newHoleStartOff`) is smaller than the
+// size of the file.  This happens when there is a hole in the file,
+// and the user is now writing data into that hole.  This function
+// moves the new block into the correct place, and rearranges all the
+// indirect pointers in the file as needed.  It returns any block
+// pointers that were dirtied in the process.
+func (fd *fileData) shiftBlocksToFillHole(
+	ctx context.Context, newTopBlock *FileBlock, newHoleStartOff int64,
+	parents []parentBlockAndChildIndex) (
+	newDirtyPtrs []BlockPointer, err error) {
+	fd.log.CDebugf(ctx, "Shifting block with offset %d for file %v into "+
+		"position", newHoleStartOff, fd.rootBlockPointer())
+
+	// Walk down the right side of the tree to find the new rightmost
+	// indirect pointer, the offset of which should match
+	// `newHoleStartOff`.  Keep swapping it with its sibling on the
+	// left until its offset would be lower than that child's offset.
+	// If there are no children to the left, continue on with the
+	// children in the cousin block to the left.  If we swap a child
+	// between cousin blocks, we must update the offset in the right
+	// cousin's parent block.  If *that* updated pointer is the
+	// leftmost pointer in its parent block, update that one as well,
+	// up to the root.
+	//
+	// We are guaranteed at least one level of indirection because
+	// `newRightBlock` should have been called before
+	// `shiftBlocksToFillHole`.
+	immedParent := parents[len(parents)-1]
+	currIndex := immedParent.childIndex
+	if off := immedParent.pblock.IPtrs[currIndex].Off; off != newHoleStartOff {
```

Comment:
Also, `newHoleStartOff` is a bit confusing, since it seems to mean the offset of where the new hole starts. But the new hole will start at `newHoleStartOff + length(this block)`, right? Maybe `newBlockStartOff`?